### ISSUE TITLE
multiprocessing Value lock sample code incorrect?

### ISSUE THREAD

**Issue by jsykes2** [python/cpython#106691]
# Documentation
Here, https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing, I find this sample code (it is about half way down the page, and I have cut out irrelevant parts)

```python
def modify(n, x, s, A):
    n.value **= 2
    x.value **= 2
    s.value = s.value.upper()
    for a in A:
        a.x **= 2
        a.y **= 2

if __name__ == '__main__':
    lock = Lock()

    n = Value('i', 7)
    x = Value(c_double, 1.0/3.0, lock=False)
    s = Array('c', b'hello world', lock=lock)
    A = Array(Point, [(1.875,-6.25), (-5.75,2.0), (2.375,9.5)], lock=lock)

    p = Process(target=modify, args=(n, x, s, A))
    p.start()
```

I believe it is incorrect. Looking at the first statement in modify(), I know that n.value **= 2 is a two step process, the use of __ipow__ followed by an assignment, effectively temp = n.value.__ipow__(2) followed by n.value = temp, or perhaps just temp = n.value.__pow__(6) followed by n.value = temp. This seems to make it clear that the RLock associated with n would not be acquired for the entire duration of the two steps, and the operation would not be safe from race conditions. A few examples that I have seen outside of the official documentation suggest that this view is correct.

Sorry to say, it is too hard to verify this experimentally. A program to demonstrate failure would have to run for a very long time before hitting on just the right conditions to produce an incorrect result.

(A clear and concise description of the issue.)

**Comment by gaogaotiantian** [comment]
I believe by "process-safe", it means the value you get will always be a full value without interruption, it does not mean you can do an arbitrary number of actions on the value - for that you always need a lock.

Integer is a little bit more hard to explain, think of an array - "process-safe" array means you have an array that when you write to it, you'll always finish before other process reads it.

When you change an array from `[1, 2, 3]` to `[4, 5, 6]`, the other process will either read `[1, 2, 3]` (before you change it), or `[4, 5, 6]` (after you change it), it will never read something like `[4, 2, 3]`.

I don't believe the docs claim that `n.value **= 2` is atomic. The piece of code itself works fine too as there's only one process spawned - no racing issue.

**Comment by jsykes2** [comment]
I agree that it doesn't explicitly say that n.value **= 2 is atomic, but this is in the section on multiprocessing, and within that it is in the subsection on shared types, and the example does explicitly create a lock and attach it to the two Arrays. And it is in a multiprocessing environment, we're only safe because the parent process immediately does a join. The vast majority of people, even programmers, are unfamiliar with race conditions and mutexes. I really do think that this example, as it stands and where it is, is certain to lead people astray.

**Comment by gaogaotiantian** [comment]
"shared types" means it can be shared between multiple processes - which is normally a more complicated job than sharing data between threads.

> we're only safe because the parent process immediately does a join

I don't think this has nothing to do with process safe, without the join, the child process may not have to time to execute.

> The vast majority of people, even programmers, are unfamiliar with race conditions and mutexes.

Well, I may agree with the conclusion, but this has nothing to do with the topic - if someone are not familiar with race conditions, they won't be able to write multi-thread or multi-process code, that's it.

`dict` (and all the built-in types I believe) is thread-safe, that does not mean you can have multiple threads do `d['a'] = d['b']` without worrying about race conditions.

May I ask what's your alternative proposal for this piece of code? Is there anything that you think would make this piece of code better?

**Comment by jsykes2** [comment]
I think we might not be talking about exactly the same thing at the moment.

My main point is that this code is in a section of the documentation that is all about multiprocessing and shared memory and locks. It creates a lock and associates it with some shared memory, and it creates a subprocess to work on that shared memory. I think that many readers are certain to expect that it will be showing them the safe way to do all those things. 

The next part I disagree with. I am very familiar with race conditions and shared memory. I came to this section to learn how Python handles it, and did indeed think that this was telling me the safe way to handle shared memory. My first thought was that Python must be a lot cleverer in the way it handles locks than I had imagined. It took a while to fully convince myself that it isn't.

I agree with what you say about a dict, but don't really see how it fits in with this. But I won't worry about that, I can see it is just an example.

Yes. My proposal is that in this example, the subprocess should acquire the appropriate locks before doing the update operations on shared memory variables. And release them afterwards of course. Then nobody could be led astray.

**Comment by gaogaotiantian** [comment]
Would you mind sharing a draft example? Maybe not for all of the values, just one of them. Let's see if it makes the example better and decide whether to move forward for this matter.

**Comment by jsykes2** [comment]
Sorry for the delay, I've been a bit overwhelmed at work recently.

Yes. That with is exactly what I had in mind.

**Comment by gaogaotiantian** [comment]
Okay, I can see there might be some benefits doing this, but I'm also not super against the original docs (because of the reasons I mentioned above). If we do something like this, I think we should throw in some comments about why the lock is acquired.

Maybe @CAM-Gerlach can share some thoughts on this?

**Comment by jsykes2** [comment]
Can't disagree with that.
As it's right after the section on locks, it shouldn't require a very big or complex explanation.
